Fix pseudo heartbeating when HB file missing

NickLaMuro opened the pull request:
With the previous version of this code, the validate_heartbeat code:

```ruby
def validate_heartbeat(w)
  last_heartbeat = workers_last_heartbeat(w)

  if w.last_heartbeat.nil?
    last_heartbeat ||= Time.now.utc
    w.update_attributes(:last_heartbeat => last_heartbeat)
  elsif !last_heartbeat.nil? && last_heartbeat > w.last_heartbeat
    w.update_attributes(:last_heartbeat => last_heartbeat)
  end
end
```

Would update the `last_heartbeat` value for the worker every time `validate_heartbeat` was called, which is not the intent.

When working correctly, the validate_heartbeat code should:

  * If no value is set in the DB, initialize it to either the time pulled from workers_last_heartbeat, or the current time (this should only happen the first time this method is called for this worker)
  * If workers_last_heartbeat is not nil, and it is a more resent heartbeat time, update the DB with the new value

Otherwise, assume we have the most up to date heartbeat value in the DB.

This change just returns `nil` if we don't have anything to provide from the heartbeat file.  This is fine since the `validate_heartbeat` method will take care of any initialization of the last_heartbeat value in the DB, so workers_last_heartbeat_to_file is just an existence check, plus a value if there is one.

Links
-----
* https://github.com/ManageIQ/manageiq/pull/15377

NickLaMuro commented:
@miq-bot assign @gtanzillo
@miq-bot assign @jrafanie
@miq-bot add_label bug

jrafanie commented on spec/models/miq_server/worker_management/heartbeat_spec.rb:
```diff
@@ -15,5 +15,75 @@
+
+      around do |example|
+        ENV["WORKER_HEARTBEAT_METHOD"] = "file"
+        ENV["WORKER_HEARTBEAT_FILE"]   = "/path/to/worker.hb"
```
If you need to rebase/fix a test, can you put this path into a `let` to DRY up some of the expectations below?

NickLaMuro replied:
@jrafanie I thought `lets` weren't able to be accessed within an "around" block, but I could be wrong.  But that was my rational for it not being DRY.

NickLaMuro replied:
I was wrong... 😞 

Will change.

jrafanie commented on spec/models/miq_server/worker_management/heartbeat_spec.rb:
```diff
@@ -15,5 +15,75 @@
+          expect(File).to receive(:exist?).with("/path/to/worker.hb")
+                                          .and_return(true, true)
+          expect(File).to receive(:mtime).with("/path/to/worker.hb")
+                                         .and_return(first_heartbeat, first_heartbeat+1)
```
^  👍

jrafanie commented on spec/models/miq_server/worker_management/heartbeat_spec.rb:
```diff
@@ -15,5 +15,75 @@
+          #   1. Sets the initial heartbeat
+          #   2. Doesn't update the worker's last_heartbeat value
+          #
+          # But the result should not change after the first loop
```
👍

jrafanie commented:
LGTM so far @NickLaMuro, just minor test fixups.  Nice find!

NickLaMuro commented:
@jrafanie @gtanzillo Made a few updates:

* Using `be_within(1.second).of(SOME_TIME_HERE)` to allow the tests to pass (rounding error when pulling from the DB I believe).
* Instead of `2.times do |i|`, doing things like `[0, 5].each do |i|` to space out the timings enough so that `be_within` isn't giving a false positive.
* Updated the "missing heartbeat file" context to loop 4 times, just to confirm the same time (probably overkill).
* Updated my comment for the "missing heartbeat file" case (made it more clear).

jrafanie commented on spec/models/miq_server/worker_management/heartbeat_spec.rb:
```diff
@@ -15,5 +15,75 @@
+              miq_server.validate_heartbeat(worker)
+            end
+
+            expect(worker.reload.last_heartbeat).to be_within(1.second).of(first_heartbeat + i)
```
👍
